File: app/src/main/java/com/petal/browser/view/NinjaToast.java
package com.petal.browser.view;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.util.TypedValue;
import android.view.Gravity;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

public class NinjaToast {

    public static void show(Context context, int stringResId) {
        if (context == null) return;
        show(context, context.getString(stringResId));
    }

    public static void show(Context context, String text) {
        if (context == null || text == null || text.isEmpty()) return;

        try {
            Toast toast = new Toast(context);

            // Resolve M3 Theme Colors dynamically
            TypedValue surfaceValue = new TypedValue();
            context.getTheme().resolveAttribute(com.google.android.material.R.attr.colorSurfaceContainerHighest, surfaceValue, true);
            if (surfaceValue.data == 0) {
                context.getTheme().resolveAttribute(com.google.android.material.R.attr.colorSurfaceContainerHigh, surfaceValue, true);
            }
            int backgroundColor = surfaceValue.data != 0 ? surfaceValue.data : Color.parseColor("#2B2D30");

            TypedValue onSurfaceValue = new TypedValue();
            context.getTheme().resolveAttribute(com.google.android.material.R.attr.colorOnSurface, onSurfaceValue, true);
            int textColor = onSurfaceValue.data != 0 ? onSurfaceValue.data : Color.WHITE;

            // Container layout
            LinearLayout container = new LinearLayout(context);
            container.setOrientation(LinearLayout.HORIZONTAL);
            container.setGravity(Gravity.CENTER);

            int paddingHorizontal = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 20, context.getResources().getDisplayMetrics());
            int paddingVertical = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 12, context.getResources().getDisplayMetrics());
            container.setPadding(paddingHorizontal, paddingVertical, paddingHorizontal, paddingVertical);

            // M3 Expressive Pill shape background with subtle elevation shadow
            GradientDrawable background = new GradientDrawable();
            background.setShape(GradientDrawable.RECTANGLE);
            background.setCornerRadius(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 28, context.getResources().getDisplayMetrics()));
            background.setColor(backgroundColor);
            container.setBackground(background);
            container.setElevation(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6, context.getResources().getDisplayMetrics()));

            // Text
            TextView textView = new TextView(context);
            textView.setText(text);
            textView.setTextColor(textColor);
            textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 14);
            textView.setTypeface(Typeface.create("sans-serif-medium", Typeface.NORMAL));
            textView.setGravity(Gravity.CENTER);

            container.addView(textView);

            toast.setView(container);
            toast.setDuration(Toast.LENGTH_SHORT);
            toast.show();
        } catch (Exception e) {
            Toast.makeText(context.getApplicationContext(), text, Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/petal/browser/view/NinjaToast.kt
package com.petal.browser.view

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.util.TypedValue
import android.view.Gravity
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast

object NinjaToast {

    @JvmStatic
    fun show(context: Context?, stringResId: Int) {
        if (context == null) return
        show(context, context.getString(stringResId))
    }

    @JvmStatic
    fun show(context: Context?, text: String?) {
        if (context == null || text.isNullOrEmpty()) return

        try {
            val toast = Toast(context)

            // Resolve M3 Theme Colors dynamically
            val surfaceValue = TypedValue()
            context.theme.resolveAttribute(com.google.android.material.R.attr.colorSurfaceContainerHighest, surfaceValue, true)
            if (surfaceValue.data == 0) {
                context.theme.resolveAttribute(com.google.android.material.R.attr.colorSurfaceContainerHigh, surfaceValue, true)
            }
            val backgroundColor = if (surfaceValue.data != 0) surfaceValue.data else Color.parseColor("#2B2D30")

            val onSurfaceValue = TypedValue()
            context.theme.resolveAttribute(com.google.android.material.R.attr.colorOnSurface, onSurfaceValue, true)
            val textColor = if (onSurfaceValue.data != 0) onSurfaceValue.data else Color.WHITE

            // Container layout
            val container = LinearLayout(context).apply {
                orientation = LinearLayout.HORIZONTAL
                gravity = Gravity.CENTER
            }

            val paddingHorizontal = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 20f, context.resources.displayMetrics).toInt()
            val paddingVertical = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 12f, context.resources.displayMetrics).toInt()
            container.setPadding(paddingHorizontal, paddingVertical, paddingHorizontal, paddingVertical)

            // M3 Expressive Pill shape background with subtle elevation shadow
            val background = GradientDrawable().apply {
                shape = GradientDrawable.RECTANGLE
                cornerRadius = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 28f, context.resources.displayMetrics)
                setColor(backgroundColor)
            }
            container.background = background
            container.elevation = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6f, context.resources.displayMetrics)

            // Text
            val textView = TextView(context).apply {
                this.text = text
                setTextColor(textColor)
                setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f)
                typeface = Typeface.create("sans-serif-medium", Typeface.NORMAL)
            }

            container.addView(textView)

            toast.view = container
            toast.duration = Toast.LENGTH_SHORT
            toast.show()
        } catch (e: Exception) {
            // Fallback to standard system toast if custom layout fails
            Toast.makeText(context, text, Toast.LENGTH_SHORT).show()
        }
    }
}
